src/schema-convertors/internalToStandard.ts

```diff
@@ -1,12 +1,324 @@
+    type: 'object',
+    properties: {
+      $undefined: {
+        type: 'boolean',
+        const: true
+      }
+    },
+    required: ['$undefined'],
+    additionalProperties: false
+  }
+});
+
+const convertInternalType = (internalType: string) => {
+  const type = { ...InternalTypeToStandardTypeMap[internalType] };
+  if (!type) throw new Error(`Encountered unknown type: ${internalType}`);
+  return type;
+};
+
+async function allowAbort(signal?: AbortSignal) {
+  return new Promise<void>((resolve, reject) =>
+    setTimeout(() => {
+      if (signal?.aborted) return reject(signal?.reason || new Error('Operation aborted'));
+      resolve();
+    })
+  );
+}
+
+async function parseType(type: SchemaType, signal?: AbortSignal): Promise<StandardJSONSchema> {
+  await allowAbort(signal);
+  const schema: StandardJSONSchema = convertInternalType(type.bsonType);
+  switch (type.bsonType) {
+    case 'Array':
+      schema.items = await parseTypes((type as ArraySchemaType).types);
+      break;
+    case 'Document':
+      Object.assign(schema,
+        await parseFields((type as DocumentSchemaType).fields, signal)
+      );
+      break;
+  }
+
+  return schema;
+}
+
+function isPlainTypesOnly(types: StandardJSONSchema[]): types is { type: JSONSchema4TypeName }[] {
+  return types.every(definition => !!definition.type && Object.keys(definition).length === 1);
+}
+
+async function parseTypes(types: SchemaType[], signal?: AbortSignal): Promise<StandardJSONSchema> {
+  await allowAbort(signal);
+  const definedTypes = types.filter(type => type.bsonType.toLowerCase() !== 'undefined');
+  const isSingleType = definedTypes.length === 1;
+  if (isSingleType) {
+    return parseType(definedTypes[0], signal);
+  }
+  const parsedTypes = await Promise.all(definedTypes.map(type => parseType(type, signal)));
+  if (isPlainTypesOnly(parsedTypes)) {
+    return {
+      type: parsedTypes.map(({ type }) => type)
+    };
+  }
+  return {
+    anyOf: parsedTypes
+  };
+}
+
+async function parseFields(fields: DocumentSchemaType['fields'], signal?: AbortSignal): Promise<{
+  required: StandardJSONSchema['required'],
+  properties: StandardJSONSchema['properties'],
+}> {
+  const required = [];
+  const properties: StandardJSONSchema['properties'] = {};
+  for (const field of fields) {
+    if (field.probability === 1) required.push(field.name);
+    properties[field.name] = await parseTypes(field.types, signal);
+  }
+
+  return { required, properties };
+}
+
+export default async function internalSchemaToStandard(
   internalSchema: InternalSchema,
   options: {
     signal?: AbortSignal
-}): Promise<StandardJSONSchema> {
-  // TODO: COMPASS-8700
-  return Promise.resolve({} as StandardJSONSchema);
+} = {}): Promise<StandardJSONSchema> {
+  const { required, properties } = await parseFields(internalSchema.fields, options.signal);
+  const schema: StandardJSONSchema = {
+    type: 'object',
+    required,
+    properties,
+    $defs: RELAXED_EJSON_DEFINITIONS
```

Comment:
non-blocking NIT: do we want to include all of the definitions, even for types that aren't used?